src/NHibernate/Cache/UpdateTimestampsCache.cs

```diff
@@ -33,34 +34,50 @@ public UpdateTimestampsCache(Settings settings, IDictionary<string, string> prop
 			updateTimestamps = settings.CacheProvider.BuildCache(regionName, props);
 		}
 
-		[MethodImpl(MethodImplOptions.Synchronized)]
+		//Since v5.1
+		[Obsolete("Please use PreInvalidate(IReadOnlyCollection<string>) instead.")]
 		public void PreInvalidate(object[] spaces)
+		{
+			//Only for backwards compatibility.
+			PreInvalidate(spaces.OfType<string>().ToList());
+		}
+
+		[MethodImpl(MethodImplOptions.Synchronized)]
+		public virtual void PreInvalidate(IReadOnlyCollection<string> spaces)
```

Comment:
I do not think this class is meant to be extended. (It should likely be sealed, but it could be done only for 6.0.) Is `virtual` necessary for NSubstitute? If yes I quite dislike to have to do that for testing purpose.
Otherwise why should it be `virtual`?
(Same for the two next `virtual` methods.)

Reply:
I'm more in the "virtual should be the default" camp, but yes, it's for NSubstitute. It's very hard to create real unit tests in NHibernate, since everything is so coupled. I figured making these methods virtual was a small price to pay for coming a bit closer to a unit test. An alternative (which I explored) is to create mocks of the underlying ICache, but that's actually quite hard, since the provider is instantiated using reflection.

Reply:
Well, not as big as making something more visible just for test. So ok.

Reply:
>Well, not as big as making something more visible just for test. So ok.

@fredericDelaporte I was unsuccessfully trying to comprehend your last comment. Could you please explain?

Reply:
Switching something to virtual just for tests is not, in most cases, a big concern. (At least, far from being as big as some other "just for tests" changes we have already seen in NHibernate, and which were raising the visibility of some members.)
So let it be for this case here.

Reply:
Ok. Thanks

Reply:
Thanks for making the fixes, @hazzik